### DISCUSSION TITLE
Calculate A1 temperature for steels

### DISCUSSION THREAD

**Discussion by dyhliu** [pycalphad/pycalphad#514] [Q&A] [answered]
I couldn't find any sample codes for calculating A1 and A3 temperatures.  So I tried to write one on my own.  I was able to obtain A3 with a few iterations using the lever rule.  However, A1 was difficult to calculate due to the potential coexistence of two FCC phases (austenite and TiC).  I tried to differentiate them using the C or Fe content in each FCC phase but was unable to do that correctly.  Would someone help?  Thank you in advance.

Daniel

**Comment by mfrichtl** [comment, marked as answer]
Daniel,

This is a common issue with calphad modeling of alloy steels where the Ae1 temperature prediction can be very low. I'll note that the Ae1 also doesn't have much practical meaning within the context of steels containing an appreciable amount of austenite stabilizing elements.

Ti is not very soluble in austenite, so you should be able to look at compositions, if your TDB file is correct. If the only other FCC phase is TiC, then you should be able simply select the one that does not contain equiatomic Ti and C, assuming you're modeling TiC as a line compound, or somewhat close to equiatomic if you're using a common assessment for the C-TI system that includes vacancies in one or both sublattices.

If that won't work for you, then would you mind posting more information like your composition of interest and sample code?

**Reply by mfrichtl** [reply]
I should've added that the A1 is going to be temperature at which the only stable FCC_A1 phase is the TiC, assuming that's the only other FCC phase in your system. Depending on your alloy composition and TDB definitions, this temperature may not actually exist in your model.

**Reply by dyhliu** [reply]
Thank you so much for your reply.  Below is my code:
from pycalphad import Database, equilibrium, variables as v
from pycalphad.plot.utils import phase_legend
# from numpy import isnan, seterr, sqrt, exp, log as ln
import numpy as np
# Importing seterr from numpy
from numpy import seterr

# Constants
T0 = 273.15  # wikipedia.org
R = 8.314462618
M_C = 12.011
M_Si = 28.086
M_Mn = 54.938
M_Cr = 51.996
M_Mo = 95.94
M_Al = 26.982
M_Nb = 92.906
M_Fe = 55.847
seterr(divide='ignore', invalid='ignore')

# Enter steel composition in wt%
xC = 0.226
xMn = 2.352
xSi = 0.52
xCr = 0
xMo = 0
xAl = 0.796
xNb = 0.02
xFe = 100 - (xC + xMn + xSi + xCr + xMo + xAl + xNb)
xC /= M_C
xMn /= M_Mn
xSi /= M_Si
xCr /= M_Cr
xMo /= M_Mo
xAl /= M_Al
xNb /= M_Nb
xFe /= M_Fe
xTotal = xC + xMn + xSi + xCr + xMo + xAl + xNb + xFe

# dbf = Database('Fe-mc-sub2060.tdb')
dbf = Database('Fe-mc-sub2060.tdb')
# phases = dbf.phases.keys()  # Including all phases is very slow.
phases = ['FCC_A1', 'BCC_A2', 'CEMENTITE ', 'M23C6']  # The fewer the phases, the faster the speed.
# elems = ['C', 'SI', 'MN', 'AL', 'CR', 'MO', 'NB', 'FE', 'VA']
elems = ['C', 'SI', 'MN', 'AL', 'NB', 'FE', 'VA']
# elems = ['C', 'SI', 'MN', 'AL', 'FE', 'VA']
conds = {  # composition converted from wt% in Phase Transformation.xlsx
    # v.T: (650+T0, 900+T0),
    v.P: 101325,
    v.X('C'): xC / xTotal,
    v.X('SI'): xSi / xTotal,
    v.X('MN'): xMn / xTotal,
    # v.X('CR'): xCr/xTotal,
    # v.X('MO'): xMo/xTotal,
    v.X('AL'): xAl / xTotal,
    v.X('NB'): xNb / xTotal
}

loT = 670 + T0
dictT = {v.T: loT}
conds.update(dictT)
res = equilibrium(dbf, elems, phases, conds)
res1 = res.NP.where(res.Phase == 'FCC_A1', drop=True).sel()
print(res1)

phases_in_eq = sorted(np.intersect1d(phases, res.Phase.values, assume_unique=False, return_indices=False))
print('Phases in equilibrium:')
print(phases_in_eq)
print(' ')
print(res.Phase.values)
print(' ')

print(res1['X_C'].values)

Below is the outputs:
"C:\Users\dyhli\My Drive\Coding\Python\PySteel\venv\Scripts\python.exe" "C:\Users\dyhli\My Drive\Coding\Python\PySteel\eqT.py" 
<xarray.DataArray 'NP' (N: 1, P: 1, T: 1, X_AL: 1, X_C: 1, X_MN: 1, X_NB: 1,
                        X_SI: 1, vertex: 2)> Size: 16B
array([[[[[[[[[2.61873956e-01, 2.36691034e-04]]]]]]]]])
Coordinates:
  * N        (N) float64 8B 1.0
  * P        (P) float64 8B 1.013e+05
  * T        (T) float64 8B 943.1
  * X_AL     (X_AL) float64 8B 0.01612
  * X_C      (X_C) float64 8B 0.01028
  * X_MN     (X_MN) float64 8B 0.02339
  * X_NB     (X_NB) float64 8B 0.0001176
  * X_SI     (X_SI) float64 8B 0.01012
  * vertex   (vertex) int32 8B 1 2
Phases in equilibrium:
['BCC_A2', 'FCC_A1']
 
[[[[[[[[['BCC_A2' 'FCC_A1' 'FCC_A1' '' '' '' '']]]]]]]]]
 
[0.01027987]

Process finished with exit code 0

As you may see, I got two FCC_A1 phases (austenite and NbC).  How may I obtain the X_C in each phase?  Thank you!

**Reply by mfrichtl** [reply]
One method is to use the vertex numbers to select each phase and examine the C content of each. A command like `res.sel(vertex=1).where(res.component=='C').X` will give you the C content of the first FCC_A1 phase. You can iterate through each vertex. The one with the lower C and Nb content should be the austenite phase.

**Reply by dyhliu** [reply]
Thank you so much for your help!
